sns_ik_examples/src/ik_tests.cpp

```diff
@@ -186,78 +203,138 @@ void test(ros::NodeHandle& nh, double num_samples_pos, double num_samples_vel,
+          q_nullspace(j)=fRand(ll(j), ul(j));
+        }
+      }
+      NullSpaceBias.push_back(q_nullspace);
+    }
   }
 
   boost::posix_time::ptime start_time;
   boost::posix_time::time_duration diff;
 
   KDL::JntArray result;
+  KDL::JntArray ns_result;
   KDL::Frame end_effector_pose;
-  int rc;
-
+  KDL::Frame end_effector_pose_check;
+  int rc, ns_rc;
+  int both_success_cnt=0;
+  bool both_success=false;
   double total_time=0;
+  double ns_total_time=0;
+  double elapsed = 0;
   uint success=0;
+  uint ns_success=0;
+  double total_ns_l2_norm_ratio=0.0;
+
   std::vector<double> kdlPos_indivTime;
 
   ROS_INFO_STREAM("*** Testing KDL with "<<num_samples_pos<<" random samples");
 
   for (uint i=0; i < num_samples_pos; i++) {
+    elapsed = 0;
+    both_success=false;
+
+    // Solve End effector pose for both natural and nullspace
     fk_solver.JntToCart(JointList[i], end_effector_pose);
-    double elapsed = 0;
+
     result = JointSeed[i];
-    start_time = boost::posix_time::microsec_clock::local_time();
     int cnt = 0;  // keep track of iteration count to enforce max number of iterations
+    start_time = boost::posix_time::microsec_clock::local_time();
     do {
       q = result; // when iterating start with last solution
-      if (nullspace_bias_task)
-        rc = kdl_nso_solver.CartToJnt(q, end_effector_pose, result);
-      else
-        rc = kdl_solver.CartToJnt(q, end_effector_pose, result);
-
+      rc = kdl_solver.CartToJnt(q, end_effector_pose, result);
       diff = boost::posix_time::microsec_clock::local_time() - start_time;
       elapsed = diff.total_nanoseconds() / 1e9;
     } while (rc < 0 && elapsed < timeout && cnt++ < 100);
     total_time += elapsed;
     kdlPos_indivTime.push_back(elapsed);
-    if (rc>=0)
+    fk_solver.JntToCart(result, end_effector_pose_check);
+    bool inPosBounds = in_pos_bounds(result, ll, ul);
+    if (rc>=0 && inPosBounds && Equal(end_effector_pose, end_effector_pose_check, 1e-3)) {
       success++;
+      both_success=true;
+    }
+
+    // Compare previous Inverse Kinematics calls against that use the nullspace
+    if(use_nullspace_bias_task) {
+      cnt = 0;
+      elapsed = 0;
+      start_time = boost::posix_time::microsec_clock::local_time();
+      ns_result = JointSeed[i];
+      do {
+        q = ns_result; // when iterating start with last solution
+        vik_nso_solver.setOptPos(NullSpaceBias[i]);
+        ns_rc = kdl_nso_solver.CartToJnt(q, end_effector_pose, ns_result);
+        diff = boost::posix_time::microsec_clock::local_time() - start_time;
+        elapsed = diff.total_nanoseconds() / 1e9;
+      } while (ns_rc < 0 && elapsed < timeout && cnt++ < 100);
+      ns_total_time += elapsed;
+      fk_solver.JntToCart(ns_result, end_effector_pose_check);
+      if (ns_rc>=0 && in_pos_bounds(ns_result, ll, ul)
+                && Equal(end_effector_pose, end_effector_pose_check, 1e-3)) {
+        ns_success++;
+        both_success &= true;
```

Comment:
If this NS task fails but the normal IK was successful, nothing sets both_success to false.